File: 2021_spring/coding_meetup_03/spiral_challenge/spiral.py
# @author Benjamin Alcocer

import sys
from bisect import bisect_left

def solve(row, col, pos):
    # Get insertion point for pos in squareArr
    index = bisect_left(squareArr, pos) 
    
    # Update row and col to the perfect square location
    row += index;
    col = row;
    
    # if position is not a perfect square, locate position on current spiral.
    # Else print perfect square location
    if squareArr[index] != pos:
        oneLess = 2 * index
        location = (squareArr[index] - position) // oneLess
        remainder = (squareArr[index] - position) % oneLess
        
        if location == 0:
            row -= remainder
        elif location == 1:
            row -= oneLess
            col -= remainder
        elif location == 2:
            row = row - (oneLess - remainder)
            col -= oneLess
        else:
            col = col - (oneLess - remainder)
    print('Line = {}, column = {}.'.format(row, col))

# Create array based off upright diagonal movement
# [1, 9, 25, 49, ...]
MAX = 50000 # n / 2
squareArr = [0] * MAX

squareValue = 1;
for i in range(MAX):
    squareArr[i] = squareValue * squareValue;
    squareValue += 2;

# For each line in stdin, solve for size, position
# size // 2 + 1 is the mid point of the matrix for row and col
for line in sys.stdin:
    size, position = map(int, line.split())
    if size != 0:
        solve(size // 2 + 1, size // 2 + 1, position)

File: 2021_spring/coding_meetup_03/spiral_challenge/spiral.cpp
/// @author Benjamin Alcocer

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

#define MAX 50000
#define ll long long
vector<ll> lookupTable(MAX);

void init()
{
    ll sq = 1;
    for (int i = 0; i < MAX; ++i)
    {
        lookupTable[i] = sq * sq;
        sq += 2;
    }
}

void solve(ll row, ll pos)
{
    auto search = lower_bound(lookupTable.begin(), lookupTable.end(), pos);
    int index = distance(lookupTable.begin(), search);
    row += index;
    ll col = row;
    if (lookupTable[index] != pos)
    {
        ll nMinusOne = 2 * index;
        ll location = (lookupTable[index] - pos) / nMinusOne;
        ll remainder = (lookupTable[index] - pos) % nMinusOne;
        
        if (location == 0) // right
        {
            row -= remainder;
        }
        else if (location == 1) // bottom
        {
            row -= nMinusOne;
            col -= remainder;
        }
        else if (location == 2) // left
        {
            row = row - (nMinusOne - remainder);
            col -= nMinusOne;
        }
        else // top
        {
            col = col - (nMinusOne - remainder);
        }
    }
    printf("Line = %lld, column = %lld.\n", row, col);
}


int main()
{
    init();

    ll size, position;
    cin >> size >> position;
    
    while (size != 0)
    {
        // solve
        solve(size / 2 + 1, position);

        cin >> size >> position;
    }
    

    return 0;
}
